### PR TITLE
Fix use_react_native to support absolute paths

### PR DESCRIPTION
**Pull request by gabrieldonadel**
## Summary:

While setting up a monorepo that required a custom react-native path location (react-native-macos in my case) I was getting the following error when running `pod install`

image

That's because `build_codegen` and `checkAndGenerateEmptyThirdPartyProvider` functions don't check if the given `react_native_path` is absolute or relative. 

This PR fixes this problem by checking if `react_native_path` starts with `/`

## Changelog:

[IOS] [FIXED] - Fix `use_react_native` to support custom react native absolute paths

## Test Plan:

Modify [rn-tester/Podfile](https://github.com/facebook/react-native/tree/main/packages/rn-tester/Podfile) to use an absolute path when calling `use_react_native` 

E.g.

```rb
rn_path = File.dirname(`node --print "require.resolve('react-native/package.json')"`) 

use_react_native!(
  path: rn_path, 
  ...
)
```
 
then run `pod install`

### CODE REVIEW

File: packages/react-native/scripts/cocoapods/codegen.rb [hunk 1]
```diff
@@ -11,8 +11,9 @@
 # - dir_manager: a class that implements the `Dir` interface. Defaults to `Dir`, the Dependency can be injected for testing purposes.
 # @throws an error if it could not find the codegen folder.
 def build_codegen!(react_native_path, relative_installation_root, dir_manager: Dir)
-    codegen_repo_path = "#{relative_installation_root}/#{react_native_path}/../react-native-codegen";
-    codegen_npm_path = "#{relative_installation_root}/#{react_native_path}/../@react-native/codegen";
+    is_react_native_path_absolute = react_native_path.start_with?("/")
+    codegen_repo_path = "#{is_react_native_path_absolute ? '' : relative_installation_root.to_s + '/'}#{react_native_path}/../react-native-codegen";
+    codegen_npm_path = "#{is_react_native_path_absolute ? '' : relative_installation_root.to_s + '/'}#{react_native_path}/../@react-native/codegen";
```
Comment (cipolleschi) [inline packages/react-native/scripts/cocoapods/codegen.rb:16]: can we change the implementation a little?
I would do something like this:

```suggestion
def basePath(react_native_path, relative_installation_root)
  if react_native_path.start_with?("/")
    react_native_path
  else 
    "#{relative_installation_root.to_s}/#{react_native_path}"
end

def build_codegen!(react_native_path, relative_installation_root, dir_manager: Dir)
  codegen_repo_path = "#{basePath(react_native_path, relative_installation_root)}/../react-native-codegen"
  codegen_npm_path = "#{basePath(react_native_path, relative_installation_root)}/../@react-native/codegen"
```

This allow us to reuse this function down the line like this:
```
'node',
        [
            "#{basePath(react_native_path, relative_installation_root)}/scripts/generate-provider-cli.js",
```
Reducing code duplication and logic duplication.

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @gabrieldonadel! Thank you so much for the contribution.

I added a suggestion to improve readability and maintainability.
Also, there should be some tests in `__tests__/codegen-test.rb`: it would be great to duplicate one of those tests with an absolute path, so we make sure not to break this in the future!

Thank you so much.

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Hi @gabrieldonadel! Thank you so much for the contribution.

I added a suggestion to improve readability and maintainability.
Also, there should be some tests in `__tests__/codegen-test.rb`: it would be great to duplicate one of those tests with an absolute path, so we make sure not to break this in the future!

Thank you so much.

**Comment by gabrieldonadel** [comment]
Hi @cipolleschi, thanks for the feedback, I've just pushed a commit updating the code to use the `basePath ` function along with a new test case on `codegen-test.rb` ensuring that absolute paths work

File: packages/react-native/scripts/cocoapods/codegen.rb (hunk 1 above)
Comment (gabrieldonadel) [reply packages/react-native/scripts/cocoapods/codegen.rb]: updated

**Comment by gabrieldonadel** [comment]
Hi @cipolleschi, is there something else that needs to be done here?

**Comment by cipolleschi** [comment]
No, sorry... I've been caught up in the release of RC.4 and had little time to followup with the PRs. Importing it now.

**Comment by NickGerleman** [comment]
Could we use something like [`File.expand_path`](https://www.rubydoc.info/stdlib/core/File.expand_path) instead of rolling this ourselves?

File: packages/react-native/scripts/cocoapods/codegen.rb [hunk 2]
```diff
@@ -108,3 +108,11 @@ def run_codegen!(
     codegen_utils.generate_react_codegen_podspec!(react_codegen_spec, codegen_output_dir)
   end
 end
+
+def basePath(react_native_path, relative_installation_root)
+  if react_native_path.start_with?("/")
```
Comment (NickGerleman) [inline packages/react-native/scripts/cocoapods/codegen.rb:113]: Another reason for reusing a standard library function, this check will not work for Windows paths.

Comment (cipolleschi) [inline packages/react-native/scripts/cocoapods/codegen.rb:113]:
```suggestion 
  if Pathname.new(react_native_path).absolute
```

This suggestion also need a `require 'pathname'` after the license.
(See [here](https://stackoverflow.com/a/2316795) for reference)

File: packages/react-native/scripts/cocoapods/codegen.rb [hunk 3]
```diff
@@ -11,8 +11,8 @@
 # - dir_manager: a class that implements the `Dir` interface. Defaults to `Dir`, the Dependency can be injected for testing purposes.
 # @throws an error if it could not find the codegen folder.
 def build_codegen!(react_native_path, relative_installation_root, dir_manager: Dir)
-    codegen_repo_path = "#{relative_installation_root}/#{react_native_path}/../react-native-codegen";
-    codegen_npm_path = "#{relative_installation_root}/#{react_native_path}/../@react-native/codegen";
+    codegen_repo_path = "#{basePath(react_native_path, relative_installation_root)}/../react-native-codegen";
```
Comment (cipolleschi) [inline packages/react-native/scripts/cocoapods/codegen.rb:14]: Following up with @NickGerleman suggestion, probably it is better to use some system APIs...
```suggestion
    codegen_repo_path = File.join(basePath(react_native_path, relative_installation_root), "..", "react-native-codegen")
```

File: packages/react-native/scripts/cocoapods/codegen.rb [hunk 4]
```diff
@@ -108,3 +108,11 @@ def run_codegen!(
     codegen_utils.generate_react_codegen_podspec!(react_codegen_spec, codegen_output_dir)
   end
 end
+
+def basePath(react_native_path, relative_installation_root)
+  if react_native_path.start_with?("/")
+    react_native_path
+  else
+    "#{relative_installation_root.to_s}/#{react_native_path}"
```
Comment (cipolleschi) [inline packages/react-native/scripts/cocoapods/codegen.rb:116]:
```suggestion
    File.join(relative_installation_root.to_s, react_native_path)"
```

**Review by cipolleschi — CHANGES_REQUESTED** [review]
Following @NickGerleman right suggestion, I left some comments to improve the codebase! :D 

Thank you both for the review and the effort!

File: packages/react-native/scripts/cocoapods/codegen.rb (hunk 3 above)
Comment (gabrieldonadel) [reply packages/react-native/scripts/cocoapods/codegen.rb:14]: Changing this seems to break some codegen tests, I'll try to investigate what's going on latter today

**Comment by gabrieldonadel** [comment]
Hi @cipolleschi, sorry for the delay on this, it seems that these changes to use system APIs kinda break all tests involving paths, what should we do about it? Should I update the tests or should we revert to using the original implementation? I'm afraid that using system APIs could break something else as it affected a bunch of tests

**Comment by cipolleschi** [comment]
So... the problem with these tests is that they are using a Mock for Pathname that can be found here: https://github.com/facebook/react-native/blob/main/packages/react-native/scripts/cocoapods/__tests__/test_utils/PathnameMock.rb

So, we have two options: 
1. enhance the PathnameMock to support `absolute` 
2. we revert to the original version

In theory, using system API should be the right way to go.
I'd say that we can timebox that to 1 hr, for example, to try and fix the tests. If we can't let's revert and I'll try that out myself once it landed. How do you feel about this?

**Comment by gabrieldonadel** [comment]
Oops, sorry for taking so long I this, I had this on my todo list for a while but couldn't get to work on that yet 😅

**Comment by cipolleschi** [comment]
no problem! with the approaching of the end of the half I ended up with some free time on my hands and I was clearing up my queue of open things. It took actually very little effort! ;)

**Pull request closed without merging**